x/clp/keeper.go

```diff
@@ -17,77 +19,90 @@ type Keeper struct {
-	return string(valueBytes)
-}
-
-// Implements sdk.AccountMapper.
-func (k Keeper) setTest(ctx sdk.Context, newTestValue string) {
-	store := ctx.KVStore(k.storeKey)
-	store.Set(testKey, []byte(newTestValue))
+	return Keeper{key, baseCoinTicker, bankKeeper, codespace, cdc}
 }
 
 // InitGenesis - store the genesis trend
-func InitGenesis(ctx sdk.Context, k Keeper, data Genesis) error {
-	k.setTest(ctx, data.Test)
+func InitGenesis(ctx sdk.Context, k Keeper, data types.Genesis) error {
 	return nil
 }
 
 // WriteGenesis - output the genesis trend
-func WriteGenesis(ctx sdk.Context, k Keeper) Genesis {
-	test := k.GetTest(ctx)
-	return Genesis{test}
+func WriteGenesis(ctx sdk.Context, k Keeper) types.Genesis {
+	return types.Genesis{}
 }
 
-// GetTest - returns the current test
-func (k Keeper) GetCLP(ctx sdk.Context, ticker string) string {
+// GetCLP - returns the clp
+func (k Keeper) GetCLP(ctx sdk.Context, ticker string) *types.CLP {
 	store := ctx.KVStore(k.storeKey)
 	valueBytes := store.Get(MakeCLPStoreKey(ticker))
-	return string(valueBytes)
+
+	clp := new(types.CLP)
+	k.cdc.UnmarshalBinary(valueBytes, &clp)
+
+	return clp
 }
 
 func (k Keeper) ensureNonexistentCLP(ctx sdk.Context, ticker string) sdk.Error {
 	clp := k.GetCLP(ctx, ticker)
-	if clp != "" {
+	if clp.Ticker != "" {
 		return ErrCLPExists(DefaultCodespace).TraceSDK("")
 	}
 	return nil
 }
 
+func (k Keeper) ensureExistentCLP(ctx sdk.Context, ticker string) sdk.Error {
+	clp := k.GetCLP(ctx, ticker)
+	if clp.Ticker == "" {
+		return ErrCLPNotExists(DefaultCodespace).TraceSDK("")
+	}
+	return nil
+}
+
 // Create CLP.
 func (k Keeper) create(ctx sdk.Context, sender sdk.AccAddress, ticker string, name string, reserveRatio int) sdk.Error {
+	if ticker == k.baseCoinTicker {
+		return ErrInvalidTickerName(DefaultCodespace).TraceSDK("")
+	}
 	err := k.ensureNonexistentCLP(ctx, ticker)
 	if err != nil {
 		return err
 	}
 	if reserveRatio <= 0 || reserveRatio > 100 {
 		return ErrInvalidReserveRatio(DefaultCodespace).TraceSDK("")
 	}
-	clp := NewCLP(sender, ticker, name, reserveRatio)
+	clp := types.NewCLP(sender, ticker, name, reserveRatio)
 	k.SetCLP(ctx, clp)
 	return nil
 }
 
+// Trade with CLP.
+func (k Keeper) tradeBase(ctx sdk.Context, sender sdk.AccAddress, ticker string, baseCoinAmount int64) sdk.Error {
+	if baseCoinAmount <= 0 {
+		return ErrNotEnoughCoins(DefaultCodespace).TraceSDK("")
+	}
+	err := k.ensureExistentCLP(ctx, ticker)
+	if err != nil {
+		return err
+	}
+	currentCoins := k.bankKeeper.GetCoins(ctx, sender)
+	currentBaseCoinAmount := currentCoins.AmountOf(k.baseCoinTicker).Int64()
+	if currentBaseCoinAmount < baseCoinAmount {
+		return ErrNotEnoughCoins(DefaultCodespace).TraceSDK("")
+	}
+	newCoins := sdk.Coins{sdk.NewCoin(ticker, baseCoinAmount)}
+	spentBaseCoins := sdk.Coins{sdk.NewCoin(k.baseCoinTicker, baseCoinAmount)}
+	finalCoins := currentCoins.Plus(newCoins).Minus(spentBaseCoins)
+	k.bankKeeper.SetCoins(ctx, sender, finalCoins)
+	return nil
```

Comment:
Not sure if I missed it – but we would need to adjust the amount of coins in the CLP too, right?